ARCHITECTURE.md

```diff
@@ -0,0 +1,57 @@
+# The Architecture of the Polarity Compiler
+
+Following the reasoning laid out by [matklad](https://matklad.github.io/2021/02/06/ARCHITECTURE.md.html), we give an overview of the architecture of the polarity compiler.
+This documentation is intended for users who want to contribute patches or features, not for users who just want to use the language.
+
+### Lexical Analysis
+
+The syntax of polarity is not whitespace sensitive and is analyzed using a generated lexer and parser.
+All code related to lexing and parsing, as well as the datatypes used to represent a parsed syntax tree, live in the `lang/parser` subdirectory as part of the `parser` crate.
+The lexer is generated with the help from the [logos](https://docs.rs/logos/latest/logos/) crate, and the specification of Tokens and the lexer lives in `lang/parser/src/lexer/mod.rs`.
+The parser is generated using the [lalrpop](https://lalrpop.github.io/lalrpop/) parser generator, and specified in `lang/parser/src/grammar/cst.lalrpop`.
+The generated parser annotates (almost) all syntax nodes with information about the source span, since this information is needed for error messages and the LSP implementation.
+
+We use a separate syntax representation for the result of parsing that we call the *concrete syntax tree (CST)*.
+The CST, which is defined in the files in  `lang/parser/src/cst/..`, should mirror the productions of the parser as closely as possible.
+We do not parse with a symbol table, so when we encounter `f` in the token stream, for example, we cannot yet tell whether `f` is a locally bound variable or a function defined at the toplevel.
+The CST therefore does not have separate constructors for variables and function calls, these must be distinguished in the lowering phase which comes after parsing.
+
+### Lowering
+
+Lowering is the process of taking the *concrete syntax tree* that is the result from parsing and turning it into an *abstract syntax tree* that is used in all later stages of the compiler.
+Here are the things that are done in the process of lowering:
+
+- The binding structure is recorded using DeBruijn indizes instead of a named representation. That is, every variable is represented using a DeBruijn index. We do not, however, forget the original names that were used by the programmer, because these names are important when we generate error messages.
```

Comment:
```suggestion
- The binding structure is recorded using DeBruijn indices instead of a named representation. That is, every variable is represented using a DeBruijn index. We do not, however, forget the original names that were used by the programmer, because these names are important when we generate error messages.
```

Reply:
Fixed